File: src/user/user.service.ts
import { Injectable, InternalServerErrorException, Logger } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import { Prisma, User } from '@prisma/client';
import * as bcrypt from 'bcrypt';
import { ErrorCodes } from 'enums/error-codes.enum';
import { PrismaService } from 'src/common/prisma/prisma.service';
import { CreateUserDto } from './dto/create-user.dto';
import { UpdateUserDto } from './dto/update-user.dto';

@Injectable()
export class UserService {
  constructor(
    private readonly configService: ConfigService,
    private readonly prisma: PrismaService,
  ) {}

  private readonly logger = new Logger(UserService.name);

  async createUser(createUserDto: CreateUserDto): Promise<void> {
    this.logger.log(`Attempting to create user with email: ${createUserDto.email}`);
    const userInDatabase = await this.prisma.user.findUnique({
      where: {
        email: createUserDto.email,
      },
    });
    if (userInDatabase) {
      this.logger.error(`User with email: ${createUserDto.email} already exists`);
      throw new InternalServerErrorException({
        error: ErrorCodes.ALREADY_EXISTS,
      });
    }
    return this.prisma.$transaction(async tx => {
      try {
        const salt = await bcrypt.genSalt(this.configService.get('SALT_ROUNDS'));
        const hash = await bcrypt.hash(createUserDto.password, salt);
        await tx.user.create({
          data: {
            email: createUserDto.email,
            name: createUserDto.name,
            password: hash,
            publicKey: createUserDto.publicKey,
            privateKey: createUserDto.privateKey,
            iv: createUserDto.iv,
          },
          select: {
            id: true,
            createdAt: true,
            updatedAt: true,
            email: true,
          },
        });
      } catch (err) {
        this.logger.error(`Failed to create user with email: ${createUserDto.email}. Error: ${err}`);
        throw new InternalServerErrorException({ error: ErrorCodes.API_ERROR });
      }
    });
  }

  async findUserById(id: string): Promise<
    Prisma.UserGetPayload<{
      select: {
        id: true;
        createdAt: true;
        updatedAt: true;
        email: true;
        name: true;
        bio: true;
        publicKey: true;
        privateKey: true;
        iv: true;
      };
    }>
  > {
    this.logger.log(`Attempting to find user with id: ${id}`);
    const user = await this.prisma.user.findUnique({
      where: { id },
      select: {
        id: true,
        createdAt: true,
        updatedAt: true,
        email: true,
        name: true,
        bio: true,
        publicKey: true,
        privateKey: true,
        iv: true,
      },
    });
    if (!user) {
      this.logger.error(`User with id: ${id} not found`);
      throw new InternalServerErrorException({ error: ErrorCodes.NOT_FOUND });
    }
    return user;
  }

  async findUserByEmail(email: string): Promise<
    Prisma.UserGetPayload<{
      select: {
        id: true;
        createdAt: true;
        updatedAt: true;
        email: true;
        password: true;
        name: true;
        bio: true;
        publicKey: true;
        privateKey: true;
        iv: true;
      };
    }>
  > {
    this.logger.log(`Attempting to find user with email: ${email}`);
    const user = await this.prisma.user.findUnique({
      where: { email },
      select: {
        id: true,
        createdAt: true,
        updatedAt: true,
        email: true,
        password: true,
        name: true,
        bio: true,
        publicKey: true,
        privateKey: true,
        iv: true,
      },
    });
    if (!user) {
      this.logger.error(`User with email: ${email} not found`);
      throw new InternalServerErrorException({ error: ErrorCodes.NOT_FOUND });
    }
    return user;
  }

  updateUser(
    id: string,
    updateUserDto: UpdateUserDto,
  ): Promise<
    Prisma.UserGetPayload<{
      select: {
        id: true;
        createdAt: true;
        updatedAt: true;
        email: true;
        name: true;
        bio: true;
        profilePicture: true;
      };
    }>
  > {
    return this.prisma.$transaction(async tx => {
      try {
        this.logger.log(`Attempting to update user with id: ${id}`);
        const UserInDatabase = await tx.user.findUnique({
          where: {
            id,
          },
        });
        if (!UserInDatabase) {
          this.logger.error(`User with id: ${id} not found`);
          throw new InternalServerErrorException({
            error: ErrorCodes.NOT_FOUND,
          });
        }
        const user = await tx.user.update({
          where: {
            id,
          },
          data: updateUserDto,
          select: {
            id: true,
            createdAt: true,
            updatedAt: true,
            email: true,
            name: true,
            bio: true,
            profilePicture: true,
          },
        });
        return user;
      } catch (err) {
        this.logger.error(`Failed to update user with id: ${id}`);
        throw new InternalServerErrorException({ error: ErrorCodes.API_ERROR });
      }
    });
  }

  deleteUser(id: string): Promise<
    Prisma.UserGetPayload<{
      select: {
        id: true;
        createdAt: true;
        updatedAt: true;
        email: true;
      };
    }>
  > {
    return this.prisma.$transaction(async tx => {
      try {
        this.logger.log(`Attempting to delete user with id: ${id}`);
        const UserInDatabase = await tx.user.findUnique({
          where: {
            id,
          },
        });
        if (!UserInDatabase) {
          this.logger.error(`User with id: ${id} not found`);
          throw new InternalServerErrorException({
            error: ErrorCodes.NOT_FOUND,
          });
        }
        const user = await tx.user.delete({
          where: { id },
          select: {
            id: true,
            createdAt: true,
            updatedAt: true,
            email: true,
          },
        });
        return user;
      } catch (err) {
        this.logger.error(`Failed to delete user with id: ${id}`);
        throw new InternalServerErrorException({ error: ErrorCodes.API_ERROR });
      }
    });
  }

  async getProfilePicture(id: User['id']): Promise<User['profilePicture']> {
    this.logger.log(`Attempting to get profile picture for user with id: ${id}`);
    const user = await this.prisma.user.findUnique({
      where: { id },
      select: { profilePicture: true },
    });
    if (!user) {
      this.logger.error(`User with id: ${id} not found`);
      throw new InternalServerErrorException({ error: ErrorCodes.NOT_FOUND });
    }
    return user.profilePicture;
  }
}

File: src/api/user/user.service.go
package user

import (
	"errors"
	"net/http"

	"easyflow-backend/src/api"
	"easyflow-backend/src/api/auth"
	"easyflow-backend/src/api/s3"
	"easyflow-backend/src/common"
	"easyflow-backend/src/database"
	"easyflow-backend/src/enum"

	"golang.org/x/crypto/bcrypt"
	"gorm.io/gorm"
)

func CreateUser(db *gorm.DB, payload *CreateUserRequest, cfg *common.Config, logger *common.Logger) (*database.User, *api.ApiError) {
	var user database.User
	if err := db.Where("email = ?", payload.Email).First(&user).Error; err == nil {
		logger.PrintfError("User with email: %s already exists", payload.Email)
		return nil, &api.ApiError{
			Code:  http.StatusConflict,
			Error: enum.AlreadyExists,
		}
	}

	password, err := bcrypt.GenerateFromPassword([]byte(payload.Password), cfg.SaltRounds)
	if err != nil {
		logger.PrintfError("Error hashing password: %s", err)
		return nil, &api.ApiError{
			Code:  http.StatusInternalServerError,
			Error: enum.ApiError,
		}
	}

	//create a new user
	user = database.User{
		Email:      payload.Email,
		Name:       payload.Name,
		Password:   string(password),
		PublicKey:  payload.PublicKey,
		PrivateKey: payload.PrivateKey,
		Iv:         payload.Iv,
	}

	if err := db.Create(&user).Error; err != nil {
		logger.PrintfError("Error creating user: %s", err)
		return nil, &api.ApiError{
			Code:  http.StatusInternalServerError,
			Error: enum.ApiError,
		}
	}

	return &user, nil
}

func GetUserById(db *gorm.DB, jwtPayload *auth.JWTAccessTokenPayload, logger *common.Logger) (*database.User, *api.ApiError) {
	var user database.User
	if err := db.Where("id = ?", jwtPayload.UserId).First(&user).Error; err != nil {
		logger.PrintfError("Error getting user: %s", err)
		return nil, &api.ApiError{
			Code:  http.StatusInternalServerError,
			Error: enum.ApiError,
		}
	}

	logger.Printf("Successfully got user: %s", user.Id)

	return &user, nil
}

func GetUserByEmail(db *gorm.DB, email string, logger *common.Logger) (bool, *api.ApiError) {
	var user database.User
	err := db.Where("email = ?", email).First(&user).Error

	if errors.Is(err, gorm.ErrRecordNotFound) {
		logger.PrintfInfo("No user with email: %s found", err)
		return false, nil
	}

	if err != nil {
		logger.PrintfInfo("An error occured while trying to find user: %s ", err)
		return false, &api.ApiError{
			Code:  http.StatusInternalServerError,
			Error: enum.ApiError,
		}
	}

	logger.PrintfInfo("User with email: %s found", email)

	return true, nil
}

func GenerateGetProfilePictureURL(db *gorm.DB, jwtPayload *auth.JWTAccessTokenPayload, logger *common.Logger, cfg *common.Config) (*string, *api.ApiError) {
	var user database.User
	if err := db.Where("id = ?", jwtPayload.UserId).First(&user).Error; err != nil {
		logger.PrintfError("Error getting user: %s", err)
		return nil, &api.ApiError{
			Code:  http.StatusInternalServerError,
			Error: enum.NotFound,
		}
	}

	imageURL, err := s3.GenerateDownloadURL(logger, cfg, cfg.ProfilePictureBucketName, user.Id, 60*60*24*7) // 1 week expiration time
	if err != nil {
		return nil, err
	}

	logger.Printf("Successfully generated profile picture URL for user: %s", user.Id)

	return imageURL, nil
}

func GenerateUploadProfilePictureURL(db *gorm.DB, jwtPayload *auth.JWTAccessTokenPayload, logger *common.Logger, cfg *common.Config) (*string, *api.ApiError) {
	var user database.User
	if err := db.Where("id = ?", jwtPayload.UserId).First(&user).Error; err != nil {
		logger.PrintfError("Error getting user: %s", err)
		return nil, &api.ApiError{
			Code:  http.StatusInternalServerError,
			Error: enum.NotFound,
		}
	}

	uploadURL, err := s3.GenerateUploadURL(logger, cfg, cfg.ProfilePictureBucketName, user.Id, 60*60)
	if err != nil {
		logger.PrintfError("Error uploading profile picture: %s", err.Error)
		return nil, &api.ApiError{
			Code:    http.StatusInternalServerError,
			Error:   enum.ApiError,
			Details: err,
		}
	}

	logger.Printf("Successfully generated profile picture upload URL for user: %s", user.Id)

	return uploadURL, nil
}

func UpdateUser(db *gorm.DB, jwtPayload *auth.JWTAccessTokenPayload, payload *UpdateUserRequest, logger *common.Logger) (*database.User, *api.ApiError) {
	var user database.User
	if err := db.Where("id = ?", jwtPayload.UserId).First(&user).Error; err != nil {
		logger.PrintfError("Error getting user: %s", err)
		return nil, &api.ApiError{
			Code:  http.StatusInternalServerError,
			Error: enum.NotFound,
		}
	}

	if payload.Name != nil {
		user.Name = *payload.Name
	}
	if payload.Bio != nil {
		user.Bio = payload.Bio
	}

	if err := db.Update(user.Id, &user).Error; err != nil {
		logger.PrintfError("Error updating user: %s", err)
		return nil, &api.ApiError{
			Code:    http.StatusInternalServerError,
			Error:   enum.ApiError,
			Details: err,
		}
	}

	logger.Printf("Successfully updated user: %s", user.Id)

	return &user, nil
}

func DeleteUser(db *gorm.DB, jwtPayload *auth.JWTAccessTokenPayload, logger *common.Logger) *api.ApiError {
	var user database.User
	if err := db.Where("id = ?", jwtPayload.UserId).First(&user).Error; err != nil {
		logger.PrintfError("Error getting user: %s", err)
		return &api.ApiError{
			Code:  http.StatusInternalServerError,
			Error: enum.NotFound,
		}
	}

	if err := db.Delete(&user).Error; err != nil {
		logger.PrintfError("Error deleting user: %s", err)
		return &api.ApiError{
			Code:  http.StatusInternalServerError,
			Error: enum.ApiError,
		}
	}

	logger.Printf("Successfully deleted user: %s", user.Id)

	return nil
}
